### PR TITLE
Random string generation API #2113

### PR DESCRIPTION
**Pull request by havlip18**
Random string generator and unit tests

### CODE REVIEW

**Comment by havlip18** [comment]
I signed it!

**Comment by jbduncan** [comment]
Hi @havlip18, have you taken the time to read the [contribution guidelines](https://github.com/google/guava/blob/master/CONTRIBUTING.md) yet? 🙂

**Comment by ogregoire** [comment]
General remarks:

* The nomenclature is rather odd considering it's Guava.
* I should be able to provide my own `Random` instance. It's important for testing and 
* I'd use a `String` insead of a `char[]`.
* Regarding the original issue, I keep my idea that I should be able to store a generator. Your implementation doesn't do that. This is the biggest win.

**Comment by dn63111** [comment]
i signed it

File: guava/src/com/google/common/base/RandomStringUtils.java [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+        return new String(res);
+    }
+
+    /**
+     * 
+     * @param len length of the returned String
+     * @param chars String od characters to be used - if chars contains non-alphabetic non-printable characters, they may appear in the returned String
+     * @return String of length len from characters chars
+     */
+    public static String random(int len, String chars) {
```
Comment (liach) [inline guava/src/com/google/common/base/RandomStringUtils.java]: How about taking a `CharSequence`?

**Comment by dn63111** [comment]
plz recheck ##line 148 
public static String random(int len, char start, char end) {..} 
or
public static String randomNumeric(int len, char start, char end) {....}

File: guava/src/com/google/common/base/RandomStringUtils.java (hunk 1 above)
Comment (fenik17) [inline guava/src/com/google/common/base/RandomStringUtils.java]: @liach CharSequence is a interface,

**Comment by fenik17** [comment]
How about filling Appendable? For exapmple:
```java
public static void randomNumeric(int len, char start, char end, Appendable target)
```
It may be useful for Writers and StringBuilders.

File: guava/src/com/google/common/base/RandomStringUtils.java (hunk 1 above)
Comment (liach) [inline guava/src/com/google/common/base/RandomStringUtils.java]: I mean to use that for the parameter.

File: guava/src/com/google/common/base/RandomStringUtils.java [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+        Random random = new Random();
+        char[] res = new char[len];
+        char[] c = new char[chars.length()];
+        chars.getChars(0, c.length, c, 0);
+        for (int i = 0; i < res.length; i++) {
```
Comment (fenik17) [inline guava/src/com/google/common/base/RandomStringUtils.java]: i < len

File: guava/src/com/google/common/base/RandomStringUtils.java (hunk 1 above)
Comment (fenik17) [inline guava/src/com/google/common/base/RandomStringUtils.java]: @liach Ok! Agree with this. 
Also charAt() can be more effective than getChars() with short CharSequence.

File: guava/src/com/google/common/base/RandomStringUtils.java [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+     */
+    public static String randomNumeric(int len, char start, char end) {
+        if (start > end) {
+            throw new IllegalArgumentException("Start char < end char!");
+        }
+        Random random = new Random();
+        char[] res = new char[len];
+        char gap = (char) (end - start);
+        char ch;
+        for (int i = 0; i < res.length; ) {
```
Comment (fenik17) [inline guava/src/com/google/common/base/RandomStringUtils.java]: i < len

File: guava/src/com/google/common/base/RandomStringUtils.java [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+            ch = (char) (random.nextInt(gap) + start);
+            if(Character.isDigit(ch)){
```
Comment (fenik17) [inline guava/src/com/google/common/base/RandomStringUtils.java]: miss code formatting (spaces)

File: guava/src/com/google/common/base/RandomStringUtils.java [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+     * @param end last usable character
+     * @return String of length len from numeric characters (Character.isDigit == true) ranging from start to end
+     */
+    public static String randomNumeric(int len, char start, char end) {
+        if (start > end) {
+            throw new IllegalArgumentException("Start char < end char!");
+        }
+        Random random = new Random();
+        char[] res = new char[len];
+        char gap = (char) (end - start);
```
Comment (fenik17) [inline guava/src/com/google/common/base/RandomStringUtils.java]: Why not int for 'gap'?

File: guava/src/com/google/common/base/RandomStringUtils.java [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+    }
+    
+    /**
+     *
+     * @param len length of the returned String
+     * @param start first usable character
+     * @param end last usable character
+     * @return String of length len from numeric characters (Character.isDigit == true) ranging from start to end
+     */
+    public static String randomNumeric(int len, char start, char end) {
```
Comment (saikiran939) [inline guava/src/com/google/common/base/RandomStringUtils.java]: Do we need to mention any documentation about Unicode characters ? For example :

start = '\u2202' and end = '\u2500' - in this case we don't have any numeric characters in this range.

File: guava/src/com/google/common/base/RandomStringUtils.java (hunk 5 above)
Comment (liach) [inline guava/src/com/google/common/base/RandomStringUtils.java]: They are just the same, in fact.

File: guava/src/com/google/common/base/RandomStringUtils.java [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+        if (start > end) {
+            throw new IllegalArgumentException("Start char < end char!");
```
Comment (liach) [inline guava/src/com/google/common/base/RandomStringUtils.java]: `"Start char <= end char!"`

File: guava/src/com/google/common/base/RandomStringUtils.java [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+
+public class RandomStringUtils {
+    
+    private static final char ASCII_PRINT_START = 32;
+    private static final char ASCII_PRINT_END = 126;
+    private static final char ASCII_NUMERIC_START = 48;
+    private static final char ASCII_NUMERIC_END = 57;
+    
+    /**
+     *
```
Comment (liach) [inline guava/src/com/google/common/base/RandomStringUtils.java]: You need to add some documentations.

**Comment by cpovirk** [comment]
Bulk closing all pull requests that are listed as needing CLA signing. If you'd like us to look at your pull request, you'll need to sign the CLA and report back here.

If this is a false positive, I apologize. Please reopen the pull request, and we'll have a look.

**Pull request closed without merging**